optuna/storages/_journal/file.py

```diff
@@ -164,28 +164,40 @@ def __init__(self, file_path: str, lock_obj: Optional[JournalFileBaseLock] = Non
     def read_logs(self, log_number_from: int) -> List[Dict[str, Any]]:
         logs = []
         with open(self._file_path, "rb") as f:
+            # Maintain remaining_log_size to allow writing by another process while reading the log
+            remaining_log_size = os.stat(self._file_path).st_size
             log_number_start = 0
             if log_number_from in self._log_number_offset:
                 f.seek(self._log_number_offset[log_number_from])
                 log_number_start = log_number_from
+                remaining_log_size -= self._log_number_offset[log_number_from]
 
             last_decode_error = None
             for log_number, line in enumerate(f, start=log_number_start):
+                if log_number < log_number_from:
+                    continue
+
+                byte_len = len(line)
+                remaining_log_size -= byte_len
+                if remaining_log_size < 0:
+                    break
                 if last_decode_error is not None:
                     raise last_decode_error
                 if log_number + 1 not in self._log_number_offset:
-                    byte_len = len(line)
                     self._log_number_offset[log_number + 1] = (
                         self._log_number_offset[log_number] + byte_len
                     )
-                if log_number < log_number_from:
+
+                # Ensure that each line ends with line separators (\n, \r\n)
+                if not line.endswith(b"\n") and not line.endswith(b"\r\n"):
```

Comment:
A nit. The code looks equivalent to the following.

```suggestion
                if not line.endswith(b"\n"):
```